nixos/modules/services/networking/wireguard-networkd.nix

```diff
@@ -22,14 +22,16 @@ let
     ;
   inherit (lib.modules) mkIf;
   inherit (lib.options) literalExpression mkOption;
-  inherit (lib.strings) hasInfix;
+  inherit (lib.strings) hasInfix replaceStrings;
   inherit (lib.trivial) flip pipe;
 
   removeNulls = filterAttrs (_: v: v != null);
 
-  privateKeyCredential = interfaceName: "wireguard-${interfaceName}-private-key";
+  escapeCredentialName = input: replaceStrings [ "\\" ] [ "_" ] input;
```

Comment:
I was curious about the escaping requirements for credentials. `systemd.exec(5)` says:

> The ID must be a short ASCII string suitable as filename in the filesystem

A filename can contain a backslash `\` so it's not clear to me why this escaping is required. I ask for two reasons:

1. I'm wondering if this points to an upstream systemd bug handling the credential names.
2. It would be nice (if possible) to use functions like `escapeSystemdPath` instead of the custom replacement in `wireguard.nix` (but we need some sense of what escaping is required).

I don't think tracking this down is a blocker for merging, but maybe we should add a comment that this is a workaround for a bug elsewhere.

Reply:
I've spent a few hours debugging systemd, there are two sites ([in `config_parse_load_credential`](https://github.com/systemd/systemd/blob/80797bbb919b3ccde4e51b349f3ca70c1157053e/src/core/load-fragment.c#L4883) and [in `exec_context_deserialize`](https://github.com/systemd/systemd/blob/80797bbb919b3ccde4e51b349f3ca70c1157053e/src/core/execute-serialize.c#L3769), which get invoked in that order) that would need an additional flag (probably [`EXTRACT_RETAIN_ESCAPE`](https://github.com/systemd/systemd/blob/80797bbb919b3ccde4e51b349f3ca70c1157053e/src/basic/extract-word.h#L14)) to not treat substrings starting with a `\` in credential names as escape sequences.

I'm planning on submitting a patch, it is not clear if supporting this is even wanted however.